sale_order_line_reorder_fields/models/sale_order_line.py

```diff
@@ -4,23 +4,41 @@ class SaleOrderLineExtend(models.Model):
     _inherit = 'sale.order.line'
 
     call_rate = fields.Float('Call Rate')
+    updating_discount  = fields.Boolean()
+    updating_call_rate  = fields.Boolean()
     
     @api.onchange('discount')
-    def change_discount(self):
+    def _onchange_discount(self):
         for line in self:
             if line.discount:
                 line.call_rate = 100 - line.discount
-                
+                line.updating_discount = True
+
     @api.onchange('call_rate')
-    def change_call_rate(self):
+    def _onchange_call_rate(self):
         for line in self:
             if line.call_rate:
                 line.discount = 100 - line.call_rate
-                
-    @api.onchange('product_id', 'price_unit', 'product_uom', 'product_uom_qty', 'tax_id')
-    def _onchange_discount(self):
-        super(SaleOrderLineExtend, self)._onchange_discount()
-
-        if self.discount is not None:
+                line.updating_call_rate = True
+    
+    @api.onchange('product_id')
+    def onchange_product_id(self):
+        self.ensure_one()
+        if self.updating_discount or self.updating_call_rate:
+            return
+        elif self.order_id and self.order_id.pricelist_id and self.order_id.pricelist_id.discount_policy == 'without_discount' and self.order_id.pricelist_id.item_ids:
+            last_item = self.order_id.pricelist_id.item_ids[-1]
+            price_str = last_item.price
+            percent_value = float(price_str.split()[0].replace('%', ''))
+            
+            self.discount = percent_value
             self.call_rate = 100 - self.discount
-            
\ No newline at end of file
+        
+    def write(self, vals):
+        res = super(SaleOrderLineExtend, self).write(vals)
+        for line in self:
+            if line.call_rate and line.discount:
+                break
```

Comment:
nếu trong if này ko xử lý gì thì ko cần thiết.
trong for này chỉ cần:
if line.discount:
                line.call_rate = 100 - line.discount

Reply:
em mới têst xong ,bắt buộc phải có DK anh nhé 
nó chạy vào elif trc , rồi tiếp tục chạy vào if để dừng

Reply:
uh đúng phải là: if not line.call_rate and line.discount:
thôi để vậy cũng dc.

Reply:
![image](https://github.com/ezt-togawa/rtw-custom/assets/127169715/a3ee59c3-ee69-4537-acb8-f8ab20192c99)

Reply:
![Uploading image.png…]()

Reply:
![image](https://github.com/ezt-togawa/rtw-custom/assets/127169715/02d4d1fb-686f-43ca-8729-3d45676c8847)

Reply:
if line.discount:
line.call_rate = 100 - line.discount  
nó sẽ báo lỗi